glow/rate_limiter_test.go

```diff
@@ -0,0 +1,84 @@
+package glow
+
+import (
+	"testing"
+	"time"
+)
+
+func TestRateLimiterSingle(t *testing.T) {
+	rl := NewRateLimiter(1, 50*time.Millisecond)
+
+	ticker := time.NewTicker(20 * time.Millisecond)
+	defer ticker.Stop()
+
+	count := 0
+	allowed := 0
+	for _ = range ticker.C {
+		if rl.Allow() {
+			allowed++
+		}
+		count++
+		if count == 4 {
+			break
+		}
+	}
```

Comment:
I'm not sure what value the ticker adds to this test. You could just use a normal loop and get the same results, right?

Also, a note on the go memory model and execution model - there's no guarantee at all that the ticker will run at a rate of once per 20ms. The only guarantee you have is that it won't run faster than once every 20 ms. So this test is potentially flaky, especially on machines that are under load in other ways.

Reply:
It's been my experience that docker / CI will randomly freeze up for 5+ seconds at a time during disk I/O, and that can throw a lot of things off. This test itself doesn't do disk I/O so that shouldn't be a problem, but any test that will fail if the runner decides to take a break for a few seconds always makes me afraid of NDFs.

Reply:
In light of that, the way I would test this ratelimiter is to have a bunch of timed processes that send requests to the ratelimiter, with sleeps between them so that you know the minimum number of requests that should have been able to go through.

then I would grab the timestamp of when the tests completed so that you can compute the maximum number of requests that should have been able to go through based on how long it took to send all of the requests.

Then I would check that the number of requests that actually went through was between those values.

Reply:
OK. I'm probably assuming different timing than we see in Go